src/Simulation/Simulators/SparseSimulator/Native/gates.h

```diff
@@ -0,0 +1,284 @@
+#pragma once
+
+
+namespace Microsoft
+{
+    namespace Quantum
+    {
+        namespace SPARSESIMULATOR
+        {
+            // Used to track differnet kind of gates,
+            // mainly for timing summary data
+             enum OP {
+                MCPhase,
+                Phase,
+                T,
+                AdjT,
+                S,
+                AdjS,
+                H,
+                MCH,
+                X,
+                MCX,
+                Y,
+                MCY,
+                Z,
+                MCZ,
+                M,
+                Measure,
+                Exp,
+                MCExp,
+                R1,
+                MCR1,
+                Rx,
+                Rz,
+                Ry,
+                MCRx,
+                MCRz,
+                MCRy,
+                SWAP,
+                MCSWAP,
+                Assert,
+                PermuteSmall,
+                PermuteLarge,
+                Proj,
+                MCProj,
+                Allocate,
+                Release,
+                NUM_OPS // counts gate types; do not add gates after this!
+            };
+
+            // Strings for the names of the various gate types
+            const static std::string op_name(OP gate_type) {
+                switch (gate_type) {
+                    case OP::MCPhase:
```

Comment:
FYI: May be I don't know something, or maybe some particular compilers allow that, but I don't expect the fragment `OP::MCPhase` to compile successfully (as opposed to fragment `MCPhase`).
If `OP` has been declared `enum class OP` (as opposed to `enum OP`) then I would expect the `OP::MCPhase` to compile successfully.
See details in [Item 10: Prefer scoped `enum`s to unscoped `enum`s](https://moodle.ufsc.br/pluginfile.php/2377667/mod_resource/content/0/Effective_Modern_C__.pdf).

Reply:
This does compile with every C++ compiler on godbolt.org. Making it scoped is a good idea however. 'enum class' will also require explicit conversion to int if it is ever used.

Reply:
Addressed here: https://github.com/sam-jaques/qsharp-runtime/pull/1